Math2/ShapeUtils.cs

```diff
@@ -1,37 +1,147 @@
+        /// A dictionary containing the circle shapes.
+        /// </summary>
+        private static Dictionary<Tuple<float, float, float, float>, Polygon2> CircleCache = new Dictionary<Tuple<float, float, float, float>, Polygon2>();
+
+        /// <summary>
+        /// A dictionary containing the rectangle shapes.
+        /// </summary>
+        private static Dictionary<Tuple<float, float, float, float>, Polygon2> RectangleCache = new Dictionary<Tuple<float, float, float, float>, Polygon2>();
+
+        /// <summary>
+        /// A dictionary containing the convex polygon shapes.
+        /// </summary>
+        private static Dictionary<int, Polygon2> ConvexPolygonCache = new Dictionary<int, Polygon2>();
+
+        /// <summary>
+        /// Creates a convex polygon from the given texture.
+        /// </summary>
+        /// <param name="Texture"></param>
+        public static Polygon2 CreateConvexPolygon(Texture2D Texture)
+        {
+            var Key = Texture.GetHashCode();
+
+            if (ConvexPolygonCache.ContainsKey(Key))
+                return ConvexPolygonCache[Key];
+
+            var data = new uint[Texture.Width * Texture.Height];
+            Texture.GetData<uint>(data);
+
+            var Points = new List<Vector2>();
+
+            for (var i = 0; i < Texture.Width; i++)
+            for (var j = 0; j < Texture.Height; j++)
+                if (data[j * Texture.Width + i] != 0)
+                    Points.Add(new Vector2(i, j));
+
+            if (Points.Count <= 2)
+                throw new Exception("Can not create a convex hull from a line.");
+
+            int n = Points.Count, k = 0;
+            List<Vector2> H = new List<Vector2>(new Vector2[2 * n]);
+
+            Points.Sort((a, b) => a.X == b.X ? a.Y.CompareTo(b.Y) : (a.X > b.X ? 1 : -1));
+
+            for (int i = 0; i < n; ++i)
+            {
+                while (k >= 2 && cross(H[k - 2], H[k - 1], Points[i]) <= 0)
+                    k--;
+                H[k++] = Points[i];
+            }
+
+            for (int i = n - 2, t = k + 1; i >= 0; i--)
+            {
+                while (k >= t && cross(H[k - 2], H[k - 1], Points[i]) <= 0)
+                    k--;
+                H[k++] = Points[i];
+            }
+
+            Points = H.Take(k - 1).ToList();
+
+
+            return ConvexPolygonCache[Key] = new Polygon2(Points.ToArray());
+        }
+        /// <summary>
+        /// Returns the cross product of the given three vectors.
+        /// </summary>
+        /// <param name="v1">Vector 1.</param>
+        /// <param name="v2">Vector 2.</param>
+        /// <param name="v3">Vector 3.</param>
+        /// <returns></returns>
+        private static double cross(Vector2 v1, Vector2 v2, Vector2 v3)
+        {
+            return (v2.X - v1.X) * (v3.Y - v1.Y) - (v2.Y - v1.Y) * (v3.X - v1.X);
+        }
+
+
+        /// <summary>
+        /// Creates a new rectangle shape with the given width, height, x and y center.
```

Comment:
It might be wise to mention that the result is not necessarily a new rectangle (easiest change might be just to make it "Fetch a rectangle shape with the given..."